src/map.rs

```diff
@@ -63,72 +74,79 @@ pub type MapResult<T> = Result<T, MapError>;
 #[derive(Debug)]
-pub enum MapEvent {
+pub(crate) enum MapEvent {
     /// To be used when a chunk is created.
-    Created {
-        /// The map index where the chunk needs to be stored.
+    CreatedChunk {
+        /// The index of the chunk.
         index: usize,
-        // /// The Handle of the Chunk.
-        // handle: Handle<Chunk>,
-        /// The vector of `Tile`s for the `Chunk`.
-        tiles: Vec<Tile>,
-    },
-    /// If the chunk needs to be refreshed.
-    ///
-    /// # Warning
-    /// May never be used, and may be removed.
-    Refresh {
-        /// The `Handle` of the `Chunk`.
+        /// The Handle of the Chunk.
         handle: Handle<Chunk>,
     },
-    /// If the chunk had been modified.
-    Modified {
+    /// An event when a layer is created for all chunks.
+    AddedLayer {
+        z: usize,
+        kind: LayerKind,
+    },
+    MovedLayer {
+        from_z: usize,
+        to_z: usize,
+    },
+    /// An event when a layer is removed for all chunks.
+    RemovedLayer {
+        z: usize,
+    },
+    /// An event when a chunk had been modified by changing tiles.
+    ModifiedChunk {
         /// The map index where the chunk needs to be stored.
-        index: usize,
+        handle: Handle<Chunk>,
         /// The `TileSetter` that is used to set all the tiles.
         setter: TileSetter,
     },
-    /// If the chunk needs to be despawned.
-    Despawned {
+    /// An event when a chunk is spawned.
+    SpawnedChunk {
+        handle: Handle<Chunk>,
+    },
+    /// If the chunk needs to be despawned, this event is used.
+    DespawnedChunk {
         /// The `Handle` of the `Chunk`.
         handle: Handle<Chunk>,
-        /// The `Entity` that needs to be despawned.
-        entity: Entity,
+        // /// The `Entity` that needs to be despawned.
+        // entity: Entity,
     },
     /// If the chunk needs to be removed.
     ///
     /// # Warning
     /// This is destructive action! All data will be dropped and removed.
-    Removed {
+    RemovedChunk {
         /// The map index where the chunk needs to be removed.
-        index: usize,
-        /// The `Entity` that needs to be despawned.
-        entity: Entity,
+        handle: Handle<Chunk>,
+        // /// The `Entity` that needs to be despawned.
+        // entity: Entity,
     },
 }
 
 /// A TileMap which maintains chunks and its tiles within.
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
-#[derive(Debug, RenderResources)]
+#[derive(Debug)]
 pub struct TileMap {
-    #[render_resources(ignore)]
     dimensions: Vec2,
     chunk_dimensions: Vec3,
-    #[render_resources(ignore)]
     tile_dimensions: Vec2,
+    current_depth: usize,
+    layers: Vec<Option<LayerKind>>,
     #[cfg_attr(feature = "serde", serde(skip))]
-    // Should change to HashSet when merged into bevy
-    #[render_resources(ignore)]
-    chunks: Vec<Option<Handle<Chunk>>>,
+    chunks: HashMap<usize, Handle<Chunk>>,
```

Comment:
Why use usize as the key? Chunks have a natural key type of `(i32, i32)` or maybe your `Point*` type. Why not use that? I think it would simplify a lot of code. For instance, you would no longer have to explicitly track the TileMap's dimensions and you could probably just get rid of `ToIndex` and all the other index related code.

Reply:
That is changing, I just kept it like that JUST for now.